### ISSUE TITLE
Registering watch seems to trigger it one too many times

### ISSUE THREAD

**Issue by mhevery** [angular/angular.js#12983]
http://plnkr.co/edit/brHNIEoTQxNQF9eZikgJ?p=preview

It contains a template:
`<p>Hello {{log('A')}}  {{log('B')}}!</p>`

Run the above plunker and open console. I would expect the console to print: `A, B` and then `A, B` as second digest runs. Instead I see `A, A, B, B` and then `A, B` when second digest runs.

Is this a bug or intended behavior?

**Comment by frfancha** [comment]
From what I know of angualr watchers it should be A, B A, B.
As mhevery, I'm very curious to know why not.

**Comment by lgalfaso** [comment]
TL;DR
it is a bug, will post a patch in a few


Long version, things happen in the following order:
* There was an overhaul on `$interpolate`, this overhaul changed `$parse` and incorporated the concept of an interceptor
*Optimization on `$parse` landed so expressions that have filters without parameters (or the parameters are constants)  would be evaluated in 2 phases, first to evaluate the expression sans the filter evaluation and then with the filter evaluation. This also used interceptors [the second evaluation issue was added here]
* More optimizations on `$parse` landed and now expressions could be evaluated in 2 phases. One to get all the possible values that could change (lets call this *state*), the *state* was checked by `$watch` to know if an expression changed. The second to continue the evaluation (as long as this *state* is provided). This, once again, used interceptors

The last change, was supposed to fix the issue, but that made the assumption that the code always would generate the 2 phases functions, but that is not always true. If the expression is simple enough (just like the one in your case) then the 2-phase evaluations functions are not generated. In this case, if a stateless interceptor was added (just like what `$interpolate` adds) then the *state* was not used and you see the function being evaluated twice. This explains why, if you change the expression from `Hello {{log('A')}} {{log('B')}}!` to `Hello {{log('A') + ' ' + log('B')}}!`, then the repetition is not there.
